src/MagnumPlugins/UfbxImporter/UfbxImporter.cpp

```diff
@@ -0,0 +1,1162 @@
+    static_cast<void>(conf);
+
+    ufbx_load_opts opts = { };
+
+    opts.generate_missing_normals = conf.value<bool>("generateMissingNormals");
+    opts.strict = conf.value<bool>("strict");
+    opts.disable_quirks = conf.value<bool>("disableQuirks");
+    opts.load_external_files = conf.value<bool>("loadExternalFiles");
+    opts.ignore_geometry = conf.value<bool>("ignoreGeometry");
+    opts.ignore_animation = conf.value<bool>("ignoreAnimation");
+    opts.ignore_embedded = conf.value<bool>("ignoreEmbedded");
+    opts.ignore_all_content = conf.value<bool>("ignoreAllContent");
+    opts.ignore_missing_external_files = true;
+
+    /* @todo: ufbx can normalize scenes to units/coordinate systems, this is a
+      very rudimentary implementation used for testing. Not sure how exposing
+      something like this in configuration would work. Also for extra goodness
+      it depends on preserveRootNode as that's where ufbx writes the mapping.
+      Another alternative would be to expose the current scene units as some
+      custom scene fields and let users do the math. */
+    if (conf.value<bool>("normalizeUnits")) {
+        opts.target_axes = ufbx_axes_right_handed_y_up;
+        opts.target_unit_meters = 1.0f;
+    }
+
+    /* We need to split meshes by material so create a dummy ufbx_mesh_material
+       containing the whole mesh to make processing code simpler. */
+    opts.allow_null_material = true;
+
+    return opts;
+}
+
+inline Int typedId(ufbx_element *element) {
+    return element ? Int(element->typed_id) : -1;
+}
+
+inline SamplerWrapping toSamplerWrapping(ufbx_wrap_mode mode) {
+    if (mode == UFBX_WRAP_CLAMP) {
+        return SamplerWrapping::ClampToEdge;
+    } else if (mode == UFBX_WRAP_REPEAT) {
+        return SamplerWrapping::Repeat;
+    } else {
+        /* @todo: What to do about unhandled enums */
+        return SamplerWrapping::Repeat; /* LCOV_EXCL_LINE */
+    }
+}
+
+inline void logError(const char *prefix, const ufbx_error &error, ImporterFlags flags) {
+    if (flags & ImporterFlag::Verbose) {
+        char message[1024];
+        ufbx_format_error(message, sizeof(message), &error);
+        Error{Utility::Debug::Flag::NoSpace} << prefix << message;
+    } else if (error.info_length > 0) {
+        Error{Utility::Debug::Flag::NoSpace} << prefix << Containers::StringView(error.description) << ": " << Containers::StringView(error.info, error.info_length);
+    } else {
+        Error{Utility::Debug::Flag::NoSpace} << prefix << Containers::StringView(error.description);
+    }
+}
+
+inline UnsignedInt unboundedIfNegative(Int value) {
+    return value >= 0 ? UnsignedInt(value) : ~UnsignedInt{};
+}
+
+struct MaterialMapping {
+    MaterialAttributeType type;
+    MaterialAttribute attrib;
+    MaterialAttribute texture;
+    MaterialAttribute textureMatrix;
+    MaterialAttribute textureCoordinates;
+    Int colorMap;
+    Int factorMap;
+};
```

Comment:
Same as mesh or scene data, the `MaterialData` support custom attributes as well -- mainly because the new PBR properties are appearing faster than I'm able to add them to `MaterialAttributeData` :D

Internally, the attributes (including the builtin ones) are represented as strings, and all strings that don't start with an uppercase letter are free for use by custom attributes. So it'd be e.g. this in the `doMaterial()` implementation:

```cpp
arrayAppend(attributes, {"subsurfaceRadius"_s, /* value of UFBX_MATERIAL_PBR_SUBSURFACE_RADIUS */};
```

Properties of custom textures are suffixed to the custom name (so e.g. `subsurfaceColorTexture` will have a `subsurfaceColorTextureMatrix` and `subsurfaceColorTextureCoordinates`), there are no restrictions on what types the custom attributes can be (they only need to have unique names).

Not sure if this would result in any significant rethinking of the `doMaterial()` implementation, but my idea to have handling of builtin and custom attributes the same would be to ~~make use of [Trade::materialAttributeName()](https://doc.magnum.graphics/magnum/namespaceMagnum_1_1Trade.html#aa17c5605fc737667a3e41371c32895b9) for enum -> string translation~~ use strings for builtin attributes as well, they have a 1:1 mapping with the `MaterialAttribute` enum names, and then having the `MaterialMapping` struct like this:

```suggestion
struct MaterialMapping {
    MaterialAttributeType type;
    Containers::StringView attrib;
    Containers::StringView texture;
    Containers::StringView textureMatrix;
    Containers::StringView textureCoordinates;
    Int colorMap;
    Int factorMap;
};
```

If it feels like too much work, it's also fine if you want to postpone this to a subsequent PR.

For convenient testing, there's [DebugTools::CompareMaterial](https://doc.magnum.graphics/magnum/classMagnum_1_1DebugTools_1_1CompareMaterial.html) in case you didn't come across it yet.

Reply:
Thanks for the writeup, it all makes a lot more sense now!

> If it feels like too much work, it's also fine if you want to postpone this to a subsequent PR.

Nah I wasn't happy with the current `doMaterial()` and I'm just glad I can make it nicer. I feel like the custom attributes will work really well with ufbx, but there's three layers we could expose:

* `ufbx_material_pbr_map` properties: These are normalized PBR properties that are quite consistent between different exporters
* FBX `Implementation` (ie. shader) mapped values: Not always supported (old 3ds Max, Blender (before I have time to implement this), etc) and there's minor differences between materials, eg. `aiStandard` has `baseColor` and `oslStandard` uses `base_color`.
* FBX raw properties: The true wilderness, see [above](https://github.com/mosra/magnum-plugins/pull/133#discussion_r1032814747) what kind of mess it can include..

Reply:
Option 1 sounds great, yes -- that's what a plugin should ultimately do, abstract away silly minor format-specific differences. I expect that I'll be gradually changing them to `MaterialAttribute` names as I make more PBR properties builtin.

Possibly eventually the "true wilderness" as an opt-in variant (like `AssimpImporter` has with `forceRawMaterialData` to be able to hammer *at least some* PBR properties out of it). But your PBR property set is quite extensive, so having access to the raw values isn't as important I think.

Reply:
Hmm should I just expose the factors as their own (custom) properties instead of premultiplying it to the value as for example in PBR the base color factor is not really semantically a multiplier on the base color, but the *contribution of the base shading model* to the result. I haven't done much field testing on materials so I'm not sure how heavily these factors are used in practice with PBR materials.

Unfortunately on the other hand on the legacy FBX Phong model the factors are used quite extensively and even the default material of `(0.8, 0.8, 0.8)` is represented as:

```
Property: "DiffuseColor", "Color", "A",1,1,1
Property: "DiffuseFactor", "Number", "A",0.800000011920929
Property: "Diffuse", "Vector3D", "",0.800000011920929,0.800000011920929,0.800000011920929
; ^ for some reason FBX includes a premultiplied value for some properties...
```

Reply:
> contribution of the base shading model

Huh. I have no idea :sweat_smile: Ideally it should be exposed in a way that, if the `DiffuseColor` is picked from the material without the application knowing about any factor, it doesn't result in a broken (overly bright) render.

~~What about multiplying with it, and then storing the factor there as a custom property as well so if someone needs to undo it, they have the input data?~~ Eh, to keep it simple, just always multiply for now (so the legacy materials make sense), and if it becomes a problem *in practice*, we can reconsider.

Reply:
Yea the factors are the bane of my existence in the materials, I can never find a clean way of dealing with them in ufbx, as it's super confusing for the users as well.